### PR TITLE
Mark `get_type_info_static()` op class methods as hidden

### PR DESCRIPTION
**Pull request by vurusovs**
### Details:
- Each plugin linked with openvino library contains `type_info_static` symbols due default visibility. In case when one of the libraries is unloaded and app tries to get opset, it leads to segfault. So mark `get_type_info_static()` as hidden to use only one implementation exactly from openvino lib. 
- Issue may be reproduced via tests (e.g. `ov_template_func_tests`) on step https://github.com/openvinotoolkit/openvino/blob/master/src/tests/ie_test_utils/functional_test_utils/src/layer_test_utils/summary.cpp#L40 when `BWDCMP_RTTI_DECLARATION` macroses are removed/not used for some ops from opset (e.g. `AdaptiveAvgPool` from opset8)

### Tickets:
- CVS-74690

### CODE REVIEW

File: src/core/include/openvino/core/rtti.hpp [hunk 1]
```diff
@@ -24,8 +24,18 @@
...
+#    if __GNUC__ >= 4
+#        define _HIDDEN __attribute__((visibility("hidden")))
+#    else
+#        define _HIDDEN
+#    endif
+#endif
+
 #define _OPENVINO_RTTI_WITH_TYPE_VERSIONS_PARENT(TYPE_NAME, VERSION_NAME, PARENT_CLASS, OLD_VERSION) \
-    static const ::ov::DiscreteTypeInfo& get_type_info_static() {                                    \
+    _HIDDEN static const ::ov::DiscreteTypeInfo& get_type_info_static() {                            \
```
Comment (ilyachur) [inline src/core/include/openvino/core/rtti.hpp:38]: Please do the same thing in the line 15.

Comment (vurusovs) [reply src/core/include/openvino/core/rtti.hpp]: Done, thanks!

File: src/core/include/openvino/op/adaptive_avg_pool.hpp [hunk 2]
```diff
@@ -15,7 +15,6 @@ namespace v8 {
 class OPENVINO_API AdaptiveAvgPool : public Op {
 public:
     OPENVINO_OP("AdaptiveAvgPool", "opset8");
-    BWDCMP_RTTI_DECLARATION;
```
Comment (ilyachur) [inline src/core/include/openvino/op/adaptive_avg_pool.hpp:18]: Before the merge we need to revert these changes.

File: src/tests/functional/inference_engine/CMakeLists.txt [hunk 3]
```diff
@@ -58,6 +58,10 @@ if (NOT ENABLE_OV_PADDLE_FRONTEND)
     list(APPEND EXCLUDED_SOURCE_PATHS "${CMAKE_CURRENT_SOURCE_DIR}/paddle_reader")
 endif()
 
+if (NOT ENABLE_LTO)
```
Comment (ilyachur) [inline src/tests/functional/inference_engine/CMakeLists.txt:61]: Why do you add this test only with disabled LTO?

Comment (vurusovs) [reply src/tests/functional/inference_engine/CMakeLists.txt:61]: Misprint from my side, thanks!

File: src/tests/functional/inference_engine/type_info_test.cpp [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+        opsets.push_back(ov::get_opset7());
+        opsets.push_back(ov::get_opset8());
+    }
+}
+
+INSTANTIATE_TEST_SUITE_P(
+        TypeInfoTest, TypeInfoTest,
+        ::testing::Values("CPU", "HETERO:CPU", "MULTI:CPU", "AUTO:CPU",
+                          "GPU", "HETERO:GPU", "MULTI:GPU", "AUTO:GPU", "BATCH:GPU",
+                          "TEMPLATE", "HETERO:TEMPLATE", "MULTI:TEMPLATE", "AUTO:TEMPLATE"));
```
Comment (ilya-lavrenov) [inline src/tests/functional/inference_engine/type_info_test.cpp:60]: IE tests must not depend on plugins unconditionally. E.i. plugin can be turned off and tests still should be workable.

Do we need to perform model compilation here?

File: src/core/include/openvino/core/rtti.hpp [hunk 5]
```diff
@@ -6,26 +6,36 @@
...
 
+#if defined _WIN32 || defined __CYGWIN__
+#    define _OPENVINO_HIDDEN_METHOD
+#else
+#    if __GNUC__ >= 4
+#        define _OPENVINO_HIDDEN_METHOD __attribute__((visibility("hidden")))
+#    else
+#        define _OPENVINO_HIDDEN_METHOD
+#    endif
+#endif
```
Comment (ilya-lavrenov) [inline src/core/include/openvino/core/rtti.hpp:17]: I suppose we need to put these definitions there https://github.com/openvinotoolkit/openvino/blob/master/src/core/include/openvino/core/visibility.hpp#L37-L40

Comment (vurusovs) [reply src/core/include/openvino/core/rtti.hpp:17]: Thanks, will be moved

File: src/tests/functional/inference_engine/type_info_test.cpp (hunk 4 above)
Comment (vurusovs) [reply src/tests/functional/inference_engine/type_info_test.cpp:60]: Model compilation is required because it triggers opsets creation which are targets for testing

File: src/core/include/openvino/op/adaptive_avg_pool.hpp (hunk 2 above)
Comment (vurusovs) [reply src/core/include/openvino/op/adaptive_avg_pool.hpp]: Reverted

File: src/tests/functional/inference_engine/CMakeLists.txt (hunk 3 above)
Comment (vurusovs) [reply src/tests/functional/inference_engine/CMakeLists.txt]: Decided to add the test in next PR

File: src/tests/functional/inference_engine/type_info_test.cpp (hunk 4 above)
Comment (vurusovs) [reply src/tests/functional/inference_engine/type_info_test.cpp]: Decided to add the test in next PR

File: src/core/include/openvino/core/rtti.hpp (hunk 5 above)
Comment (vurusovs) [reply src/core/include/openvino/core/rtti.hpp]: Moved

**Comment by ilya-lavrenov** [comment]
How binary size is changed on Ubuntu platform?

**Comment by vurusovs** [comment]
> How binary size is changed on Ubuntu platform?
No changes

|library|master|branch|
|----|----|----|
|libopenvino_c.so | 305408 | 301312|
libopenvino_gapi_preproc.so | 1308824 | 1308824
libopenvino_gapi_preproc_s.a | 5728918 | 5729022
libopenvino_hetero_plugin.so | 366800 | 366800
libopenvino_intel_cpu_plugin.so | 33283880 | 33234728
libopenvino_intel_gna_plugin.so | 4064328 | 4060232
libopenvino_intel_gna_plugin_test_static.a | 67093782 | 67151222
libopenvino_intel_gpu_graph.a | 32002544 | 32002544
libopenvino_intel_gpu_kernels.a | 25539930 | 25539930
libopenvino_intel_gpu_plugin.so | 18988496 | 18988496
libopenvino_intel_gpu_runtime.a | 1356030 | 1356030
libopenvino_intel_myriad_plugin.so | 6095240 | 6074760
libopenvino_ir_frontend.so | 342536 | 342536
libopenvino_mock1_frontend.so | 88496 | 88496
libopenvino_onnx_frontend.so | 3777672 | 3769480
libopenvino_paddle_frontend.so | 986872 | 986872
libopenvino_template_extension.so | 80088 | 75992
libopenvino_template_plugin.so | 7764600 | 7727736
libopenvino_tensorflow_fe.so | 2719768 | 2715672
libov_protobuf_shutdown.a | 8862 | 8862
libov_shape_inference.a | 3100 | 3100
libpaddle_fe_standalone_build_test.a | 11784 | 11784
libprotobuf-lite.a | 4381172 | 4381172
libprotobuf.a | 8028148 | 8028148
libprotoc.a | 8951572 | 8951572
libpugixml.a | 1263804 | 1263804
libsharedTestClasses.a | 44964680 | 45052928
libtemplate_extension.so | 174504 | 174504
libtensorflow_fe_standalone_build_test.a | 11054 | 11054
libtest_builtin_extensions_1.so | 31152 | 31152
libtest_builtin_extensions_2.so | 18864 | 18864
libunitTestUtils.a | 189488 | 189488
libutil.a | 147296 | 147296
libvpu_common_lib.a | 63015262 | 63121662
libvpu_graph_transformer.a | 150886112 | 150900768
libzlib.a | 576170 | 576170
pcie-ma2x8x.mvcmd | 2099248 | 2099248
plugins.xml | 874 | 874
usb-ma2x8x.mvcmd | 2298632 | 2298632

**Comment by vurusovs** [comment]
@ilyachur good catch! Size of libopenvino.so decreased on ~150 Kb:
|Lib|master|branch|
|-----|-----|-----|
|libopenvino.so|13386360 |13230712|

**Review by ilyachur — APPROVED** [review]
